### PR TITLE
Overhaul Schedule Select | Edit | Delete

### LINKED ISSUE #658: Move Rename and Delete actions into Select menu
Screenshot 2023-08-20 at 4 21 55 PM

From a user's perspective, it would create a more intuitive experience to have these options visible when you open the schedule Select menu. It would also allow for other schedules to be renamed/deleted even when they're not currently selected.

### PR DESCRIPTION
**Pull request by KevinWu098**
## Summary
1. Attached rename and delete dialogs to every schedule in the `scheduleNames` array.
2. Created new on[Open|Close|Close] to manage propagation and props to the dialogs.
3. Added styling to the icons, also increasing the size of the Add Schedule button for consistency.
4. Added `scheduleIndex` as a prop to the assortment of "deleteSchedule" functions in `Schedules.ts`, `AppStore.ts`, `AppStoreActions.ts`. Previously `scheduleIndex` was based on `this.currentScheduleIndex` which would cause issues, for instance, when trying to delete Schedule 3 while having Schedule 2 selected.
5. Deleted  `EditSchedule.tsx` which contained the Rename and Delete dialogs previously. (For future reading clarity, the rename dialog is a part of `ScheduleNameDialog.tsx`, which also contains the Add Schedule dialog.

The new look:
New Look

Delete Demo:
chrome-capture-2023-7-20 (1)

Rename Demo:
chrome-capture-2023-7-20 (2)

## Test Plan
Make sure all previous functionality is as intended:
1. Swapping between schedules
2. Deleting
3. Renaming
4. Making sure that schedule indices are not mangled up and the appropriate schedule is updated or swapped to

## Issues
Closes #658

## Misc
"only potential negative is [Fred] might just overwrite it if I change it [in #654] 💀" 

## TO-DO

- [x] Refactor the deleteSchedule functions which still have remnants of their prior behavior (specifically always calling `this` or saying "the current schedule".)

### CODE REVIEW

**Review by Voark — DISMISSED** [review]
LGTM! Definitely like this layout better than our current one. Only nitpick (that isn't even your fault but might as change it because it is so in proximity to your changes) is the delete schedule dialog shouldn't have the undo warning because deleting schedules can be undone. Another nitpick is that I think we should use rem/em instead of px for the padding because it keeps the styling more consistent across devices.

**Comment by KevinWu098** [comment]
> LGTM! Definitely like this layout better than our current one. Only nitpick (that isn't even your fault but might as change it because it is so in proximity to your changes) is the delete schedule dialog shouldn't have the undo warning because deleting schedules can be undone. Another nitpick is that I think we should use rem/em instead of px for the padding because it keeps the styling more consistent across devices.

I'll get to changing that right now :)

**Comment by KevinWu098** [comment]
I only adjusted px to rem for code that I changed in this PR, but lmk if you want me to also change previously existing code in these files to rem!

**Comment by KevinWu098** [comment]
## Issue
After having stared at this for a little while, I'm noticing that there is / there will be a visual overlap between this PR and the existing "clear schedule" button. Specifically, both clear schedule and delete schedule would now use the same icon, possibly creating confusion for users.

Screenshot 2023-08-21 at 3 12 22 PM

-------------

## Possible Solutions 
The solution would be to change either the delete schedule or the clear schedule icon (which would also impact #665's intended changes)

Here are the options: *as a note, the results from searches "clear" and "erase" are either included in the "delete" search or not relevant. 

Screenshot 2023-08-21 at 3 09 26 PM

Technically the "Clear All" or "Clear" icons would make sense for **_clearing_** the schedule, but IMO the visual language isn't really conveyed well that they're "clearing." Maybe the "Clear" symbol would work (?)... 🫠 [If only there was an eraser icon](https://fonts.google.com/icons?icon.query=eraser&icon.set=Material+Symbols)

As always, would love to hear y'alls input! 💖

**Comment by Voark** [comment]
> I only adjusted px to rem for code that I changed in this PR, but lmk if you want me to also change previously existing code in these files to rem!

No need to touch extra parts of the code base IMO. 

in regards to the icon maybe we could change the delete schedule icon to like the X (maybe make it red?) or something but I have no strong opinions. Alternatively maybe we could consult some actual design people.

**Comment by KevinWu098** [comment]
> > I only adjusted px to rem for code that I changed in this PR, but lmk if you want me to also change previously existing code in these files to rem!
> 
> No need to touch extra parts of the code base IMO.
> 
> in regards to the icon maybe we could change the delete schedule icon to like the X (maybe make it red?) or something but I have no strong opinions. Alternatively maybe we could consult some actual design people.

@Voark After going through a few options and considering the available icons, I think that your idea to use a red Clear icon best balances making the most visual sense with affecting the existing user experience as little as possible.

New Look

File: apps/antalmanac/src/components/Calendar/CalendarToolbar.tsx [hunk 1]
```diff
@@ -84,25 +93,51 @@ const CalendarPaneToolbar = ({
...
     return (
         <Paper elevation={0} variant="outlined" square className={classes.toolbar}>
-            <EditSchedule scheduleNames={scheduleNames} scheduleIndex={currentScheduleIndex} />
-
             <Select
                 classes={{ root: classes.rootScheduleSelector }}
                 className={classes.scheduleSelector}
                 value={currentScheduleIndex}
-                onChange={handleScheduleChange}
+                onChange={(e) => handleScheduleChange(e)}
```
Comment (MinhxNguyen7) [inline apps/antalmanac/src/components/Calendar/CalendarToolbar.tsx]: This can just be `onChange={handleScheduleChange}`

**Review by MinhxNguyen7 — DISMISSED** [review]
I like the design decisions. I've only got nitpicks. 

I think I prefer not coloring the X, i.e., leaving it the same color as the delete button. I don't feel that strongly about it, so do what you will.

Also, the schedule selection dropdown doesn't go from the button down but also goes above, which is visually annoying. I know it's been there before, but it'd be nice if you could change that here. It should just be a change of the location relative to the anchor.

**Comment by KevinWu098** [comment]
> I like the design decisions. I've only got nitpicks.
> 
> I think I prefer not coloring the X, i.e., leaving it the same color as the delete button. I don't feel that strongly about it, so do what you will.
> 
> Also, the schedule selection dropdown doesn't go from the button down but also goes above, which is visually annoying. I know it's been there before, but it'd be nice if you could change that here. It should just be a change of the location relative to the anchor.

Done! 

I'm also not of a super strong opinion regarding the red, and I think that since the X is within the schedule Select, along with the fact that a dialog will pop up if they decide to click the X, users will be able to either intuit the meaning or just figure it out by clicking. 

TL;DR got rid of the red 👍

**Review by ap0nia — CHANGES_REQUESTED** [review]
Using a `Select` component and attempting to render extraneous content like buttons inside of its `option` elements is fundamentally wrong, i.e. per some HTML spec: don't put buttons inside a select menu's `option` because when you click on the option, it should cause that value to be selected and nothing else (source: trust me bro).

A solution that could mimic the select menu sufficiently would be the [Autocomplete component](https://mui.com/material-ui/react-autocomplete/), which **doesn't** use a select menu, but rather a `input` used to control the text being displayed, and a combination of `ul` and `li` to render the options. By using the [`renderOption` prop](https://mui.com/material-ui/api/autocomplete/#Autocomplete-prop-renderOption), the `li` itself can be customized.

An example is demonstrated below:
https://github.com/icssc/AntAlmanac/assets/66761366/8ef12383-9d24-48ae-9e34-49d3e230f097

I'll implement this on a separate branch and PR to this branch.

**Comment by Voark** [comment]
Considering how extensive this PR is getting and that I was going to redo it in my #654 maybe it is worth just implementing the changes I was going to make anyways here instead, the main change being making schedules reorderable. 

Here's a [reorderable list example](https://codesandbox.io/s/material-ui-sortable-list-with-react-smooth-dnd-swrqx?file=/src/index.js) I found for reference that I was thinking of turning the schedule selector into. @KevinWu098 do you want to add this to your PR? If not I'll just do it in mine after this gets merged.

**Comment by KevinWu098** [comment]
> Considering how extensive this PR is getting and that I was going to redo it in my #654 maybe it is worth just implementing the changes I was going to make anyways here instead, the main change being making schedules reorderable.
> 
> Here's a [reorderable list example](https://codesandbox.io/s/material-ui-sortable-list-with-react-smooth-dnd-swrqx?file=/src/index.js) I found for reference that I was thinking of turning the schedule selector into. @KevinWu098 do you want to add this to your PR? If not I'll just do it in mine after this gets merged.

We love ever-expanding PR scopes :')! I think both plans have merit, but unless the ability to reorder schedules is tied to another feature in that other ever-expanding PR of yours ( /j ), I think it would be cool to tie in the reordering into this PR! @Voark

**Comment by ap0nia** [comment]
> Considering how extensive this PR is getting and that I was going to redo it in my #654 maybe it is worth just implementing the changes I was going to make anyways here instead, the main change being making schedules reorderable.
> 
> Here's a [reorderable list example](https://codesandbox.io/s/material-ui-sortable-list-with-react-smooth-dnd-swrqx?file=/src/index.js) I found for reference that I was thinking of turning the schedule selector into. @KevinWu098 do you want to add this to your PR? If not I'll just do it in mine after this gets merged.

Let's just get this merged as-is first. Drag-and-drop re-ordering is a whole ordeal with package installation and more shenanigans.

**Comment by KevinWu098** [comment]
> > Considering how extensive this PR is getting and that I was going to redo it in my #654 maybe it is worth just implementing the changes I was going to make anyways here instead, the main change being making schedules reorderable.
> 
> > 
> 
> > Here's a [reorderable list example](https://codesandbox.io/s/material-ui-sortable-list-with-react-smooth-dnd-swrqx?file=/src/index.js) I found for reference that I was thinking of turning the schedule selector into. @KevinWu098 do you want to add this to your PR? If not I'll just do it in mine after this gets merged.
> 
> 
> 
> Let's just get this merged as-is first. Drag-and-drop re-ordering is a whole ordeal with package installation and more shenanigans.

Sounds good to me!

**Comment by Voark** [comment]
> > > Considering how extensive this PR is getting and that I was going to redo it in my #654 maybe it is worth just implementing the changes I was going to make anyways here instead, the main change being making schedules reorderable.
> > 
> > 
> > > 
> > 
> > 
> > > Here's a [reorderable list example](https://codesandbox.io/s/material-ui-sortable-list-with-react-smooth-dnd-swrqx?file=/src/index.js) I found for reference that I was thinking of turning the schedule selector into. @KevinWu098 do you want to add this to your PR? If not I'll just do it in mine after this gets merged.
> > 
> > 
> > Let's just get this merged as-is first. Drag-and-drop re-ordering is a whole ordeal with package installation and more shenanigans.
> 
> Sounds good to me!

There's nothing in my PR that is required but I did want to add it because the term selector brings up the first schedule in said term so I wanted users to be able to change the schedule shown. I also just had the realization that some sort of star thing would be more intuitive and mean I don't have to overwrite all your new changes too.

**Review by ap0nia — APPROVED** [review]
Looks good. I have to merge this first out of all the other PRs you have open @KevinWu098 because their implementations of the calendar toolbar are less complete lol
